### PR TITLE
feat: ActiveCampaign esp integration

### LINKED ISSUE #765: ActiveCampaign provider integration
As we work closer with BlueLena, it would be good to get a proper integration with ActiveCampaign instead of forcing those users to use the Manual provider. In their work so far, BlueLena has found that the Manual provider adds extra friction to the workflow which a proper ActiveCampaign integration would reduce.

[Here's the documentation for the ActiveCampaign API.](https://developers.activecampaign.com/reference/overview)

### PR DESCRIPTION
**Pull request by miguelpeixe**
### All Submissions:

* [x] Have you followed the [Newspack Contributing guideline](https://github.com/Automattic/newspack-plugin/blob/master/.github/CONTRIBUTING.md)?
* [x] Does your code follow the [WordPress' coding standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/) and [VIP Go coding standards](https://vip.wordpress.com/documentation/vip-go/code-review-blockers-warnings-notices/)?
* [x] Have you checked to ensure there aren't other open [Pull Requests](../../pulls) for the same update/change?

### Changes proposed in this Pull Request:

Implements ActiveCampaign ESP.

Closes #765.

### How to test the changes in this Pull Request:

A valid [ActiveCampaign](https://www.activecampaign.com/) account is required for testing this PR, ping me if you'd like to use my test account.

#### Basic integration

1. On a fresh install, check out this branch and run `npm run build`
2. Go through the onboarding wizard and enable "Newsletters" while on the "Services" step
3. Select ActiveCampaign, apply your account URL and key and finish the onboarding wizard
4. Visit Newsletter -> Settings and confirm your settings are applied

#### Draft campaign synchronization

1. Start a new newsletter using a layout
2. Confirm you are able to select a list from the newsletter editor sidebar
3. Enter sender information and click "Save draft"
4. Visit your ActiveCampaign dashboard and confirm the campaign exists with the following name: `Newspack Newsletters #{post-id}`

#### Send test email

1. While on the newsletter editor, enter your email address in the "Test email" section
2. Click to send and wait a few minutes to confirm you receive the test email

If you have a newly created account, you might see a "Please send this campaign through the web interface to continue.". This is an account-wide security measure from ActiveCampaign and should be resolved for any campaign after sending a number of campaigns. This has not affected sending campaigns on my tests. Reference: https://community.activecampaign.com/t/api-use-web-interface-error/10403

#### Sending campaign

1. Using the same campaign, click to Send and confirm
2. Confirm on the ActiveCampaign dashboard that the campaign is marked as sent
3. Wait a few minutes to receive the email

#### Removing trashed campaign

1. Move the sent campaign to the trash and confirm it is not removed from ActiveCampaign (only draft campaigns should be removed)
2. Create a new draft campaign, fill in the sender name, email, select a list and save the draft
3. Confirm the new draft is synced to ActiveCampaign by visiting the dashboard
4. Move the draft to the trash and confirm the campaign has been removed from ActiveCampaign

#### Adding contact

1. Use a WordPress instance properly configured to use Stripe Reader Revenue setup
2. Setup the ActiveCampaign newsletter ESP
3. On "Stripe Settings" set a list to add contacts to
4. Test the donation flow with newsletter sign up and confirm the contact is added to the list

### Other information:

* [x] Have you added an explanation of what your changes do and why you'd like us to include them?
* [ ] Have you written new tests for your changes, as applicable?
* [x] Have you successfully ran tests with your changes locally?

### CODE REVIEW

File: includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign.php [hunk 1]
```diff
@@ -0,0 +1,519 @@
...
+						'sdate'  => $schedule_date->format( 'Y-m-d H:i:s' ),
+					],
+				] 
+			);
+			if ( is_wp_error( $send_result ) ) {
+				$error = $send_result;
+			}
+		}
+
+		if ( $error ) {
```
Comment (adekbadek) [inline includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign.php:454]: `$error` might not be defined (if everything went fine), causing a `Undefined variable: error` PHP notice.

**Review by adekbadek — CHANGES_REQUESTED** [review]
After trying to save a draft newsletter, nothing happened. The response from the API was a `200` (a bit misleading!) and message: `You do not have any mailing addresses. Please create one to create a campaign.` After adding a physical address in the AC settings, the draft newsletter was saved as a campaign via the API. This error message should be visible in the UI.

File: includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign.php (hunk 1 above)
Comment (miguelpeixe) [reply includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign.php:456]: Fixed by assigning the variable earlier in 79eea75b82fa54b314f5f05d091c752dd160573c

File: includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign.php [hunk 2]
```diff
@@ -0,0 +1,519 @@
...
+	}
+
+	/**
+	 * Get campaign name.
+	 *
+	 * @param WP_Post $post Post object.
+	 * @return String Campaign name.
+	 */
+	private function get_campaign_name( $post ) {
+		return 'Newspack Newsletter #' . $post->ID;
```
[the remark this answers is not in the document]
Comment (miguelpeixe) [reply includes/service-providers/active_campaign/class-newspack-newsletters-active-campaign.php]: I agree! d14f5aad92116b5b58eb3cad334d7067c1fe6900 changes the campaign name pattern.

The API does not provide any endpoint for editing a created campaign (only for modifying its status, scheduling, deleting, etc). We must keep the campaign metadata as static and immutable as possible.

**Comment by miguelpeixe** [comment]
Thanks for the review, Adam!

> One edge-casey remark: after trying to save a draft newsletter, nothing happened. The response from the API was a 200 (a bit misleading!) and message: You do not have any mailing addresses. Please create one to create a campaign. After adding a physical address in the AC settings, the draft newsletter was saved as a campaign via the API. This error message should be visible in the UI.

The draft save action is optimistic, you can see that prior to adding a sender or selecting a list the `sync()` method is called on [retrieve](https://github.com/Automattic/newspack-newsletters/pull/775/files#diff-632bca06fd7deb73cb1e8c814bae3a6583d8474bc156751fe80c1a19cf8767a0R195-R201) and [save](https://github.com/Automattic/newspack-newsletters/pull/775/files#diff-632bca06fd7deb73cb1e8c814bae3a6583d8474bc156751fe80c1a19cf8767a0R400) but its errors are ignored and the campaign is not created.

One issue I noticed is that in any unhandled validation error, the send can be attempted but the resulting error would always be 404 and not the sync error. 2c7df7cb39f209c292cc9302373793e2caba1242 adds a sync with error handled on send if the campaign was never previously created. This ensures that the user gets the proper error message coming from ActiveCampaign.

**Pull request merged**